snuba/query/snql/parser.py

```diff
@@ -1317,4 +1395,26 @@ def parse_snql_query(
 
     # Validating
     _post_process(query, VALIDATORS)
+
+    if isinstance(query, LogicalQuery):  # avoid composite for now
+        if is_in_experiment(query, "", "snql.sort.query.projects", None):
+            # Experiment with sorting the query to improve cache hit rate
+            sorting_rate = state.get_config("snql.sort.query.rate", 0.0)
+            sorting_rate = 0.0 if not isinstance(sorting_rate, float) else sorting_rate
+
+            timer = Timer("query.sorting")
+            if settings.TESTING or random.random() < sorting_rate:
+                assert isinstance(query, LogicalQuery)
+                before = format_query(query)
+                _post_process(query, [_sort_query])
+                query.add_experiment("query-sorting", "true")
+                after = format_query(query)
+                sentry_sdk.set_tag("query-sorting.query_changed", str(before == after))
```

Comment:
Does this metric tell you a lot? It will be logged almost every time as nobody tries to sort the query now, so you would be adding a format call without gathering a lot of useful data.
I think what you would like to gather is how many queries are actually reordered when there was an equivalent version in cache. But that would be harder as you would have to find the sorted cache key and the unsorted one and see what is in cache.

Reply:
So first this is a tag not a metric, second it will only be added if we sort the query, which is 50% of certain projects.

I see where you're coming from. I wanted a more direct metric for success beyond the cache_hit rate. I added this because I thought if the sorting is doing nothing most of the time, then it's not worth all this processing just to end up with the same query.

Reply:
> So first this is a tag not a metric

My concern is with the additional format call, that scans the entire query.

What about extracting this information from the sorting method itself. For each clause you can check if that is the same expression as it was before. If at least one of these checks returns true, you do not check again for that query. Most of the times we will just check the first expression and give up.

Anyway I think a better approach to get a valid impact measure for this experiment is to sort the query right before using the cache as mentioned above. Then you can calculate the cache key of both the sorted and unsorted query and try to get both of them from cache.